File: Seccion 16 - Arboles en C++/Seccion_16_Dart/06_recorrer_postorden.dart

import 'dart:io';

class Nodo {

  int dato = null;
  Nodo izq = null;
  Nodo der = null;
  Nodo padre = null;

  Nodo(this.dato);

  String str() {
    return "$dato -> ";
  }
}

class Arbol {

  Nodo raiz = null;

  Arbol();

  void agregarNodo(Nodo nodo) {
    if (raiz == null) {
      raiz = nodo;
    } else {
      Nodo aux = this.raiz;
      Nodo padre;

      while (aux != null) {
        padre = aux;
        if (nodo.dato >= aux.dato) {
          aux = aux.der;
        } else {
          aux = aux.izq;
        }
      }

      nodo.padre = padre;
      if (nodo.dato >= padre.dato) {
        padre.der = nodo;
      } else {
        padre.izq = nodo;
      }
    }
  }

  void mostrarArbol(Nodo nodo, int cont) {
    if (nodo == null) {
      stdout.write("");
    } else {
      mostrarArbol(nodo.der, cont+1);

      for (int i = 0; i < cont; i++) {
        stdout.write("   ");
      }

      print(nodo.dato);
      mostrarArbol(nodo.izq, cont+1);
    }
  }

  bool buscarNodo(Nodo nodo, int n) {
    if (nodo == null) {
      return false;
    } else if (nodo.dato == n) {
      return true;
    } else if (n < nodo.dato) {
      return buscarNodo(nodo.izq, n);
    } else {
      return buscarNodo(nodo.der, n);
    }
  }

  String preOrden(Nodo nodo) {
    if (nodo == null) {
      return "";
    } else {
      stdout.write(nodo.str());
      preOrden(nodo.izq);
      preOrden(nodo.der);
    }
  }

  String inOrden(Nodo nodo) {
    if (nodo == null) {
      return "";
    } else {
      preOrden(nodo.izq);
      stdout.write(nodo.str());
      preOrden(nodo.der);
    }
  }

  String postOrden(Nodo nodo) {
    if (nodo == null) {
      return "";
    } else {
      preOrden(nodo.izq);
      preOrden(nodo.der);
      stdout.write(nodo.str());
    }
  }

  Nodo get_raiz() {
    return raiz;
  }
}

void menu() {
  int n;
  int opcion = 0;
  Nodo nodo;
  int contador = 0;

  Arbol arbol = new Arbol();

  do {
    stdout.writeln("\t.:MENU:.");
    stdout.writeln("1. Insertar un nuevo nodo.");
    stdout.writeln("2. Mostrar el árbol completo.");
    stdout.writeln("3. Buscar un elemento en el árbol.");
    stdout.writeln("4. Recorrer el árbol en PreOrden.");
    stdout.writeln("5. Recorrer el árbol en InOrden.");
    stdout.writeln("6. Recorrer el árbol en PostOrden.");
    stdout.writeln("7. Salir.");
    stdout.write("Opción: ");
    opcion = int.parse(stdin.readLineSync());

    switch (opcion) {
      case 1:
        stdout.write("\nDigite un número entero: ");
        n = int.parse(stdin.readLineSync());
        nodo = new Nodo(n);
        arbol.agregarNodo(nodo);
        break;
      case 2:
        stdout.writeln("\nMostrando el árbol completo:");
        arbol.mostrarArbol(arbol.get_raiz(), contador);
        break;
      case 3:
        stdout.write("\nDigite el elemento a buscar: ");
        n = int.parse(stdin.readLineSync());
        if (arbol.buscarNodo(arbol.get_raiz(), n)) {
          print("El elemento $n SÍ se encuentra en el árbol.");
        } else {
          print("El elemento $n NO se encuentra en el árbol.");
        }
        break;
      case 4:
        print("\nRecorrido del árbol en PreOrden:");
        arbol.preOrden(arbol.get_raiz());
        print("");
        break;
      case 5:
        print("\nRecorrido del árbol en InOrden:");
        arbol.inOrden(arbol.get_raiz());
        print("");
        break;
      case 6:
        print("\nRecorrido del árbol en PostOrden:");
        arbol.postOrden(arbol.get_raiz());
        print("");
        break;
      
      default:
        break;
    }
  } while (opcion != 7);
}

void main() {

  menu();

}

File: Seccion 16 - Arboles en C++/Seccion_16_Python/06_recorrer_postorden.py

class Nodo:
    
    def __init__(self):
        self.izq = None
        self.der = None
        self.dato = None
        
# Función para crear un nuevo nodo
def crear_nodo(n: int):
    nuevo_nodo = Nodo()
    
    nuevo_nodo.dato = n
    nuevo_nodo.izq = Nodo()
    nuevo_nodo.der = Nodo()
    
    return nuevo_nodo

# Función para insertar elementos al arbol
def insertar_nodo(arbol: Nodo, n: int):
    if arbol.dato == None: # Si el árbol está vacío
        arbol = crear_nodo(n)
    else: # Si el árbol tiene un nodo o más
        valor_raiz = arbol.dato # Obtenemos el valor raiz
        
        if n < valor_raiz:
            arbol.izq = insertar_nodo(arbol.izq, n) # Si el elemento es menor que la raiz
        else:
            arbol.der = insertar_nodo(arbol.der, n) # Si el elemento es mayor que la raiz
    return arbol

def mostrar_arbol(arbol: Nodo, cont: int):
    if arbol.dato == None: # Si el arbol está vacío
        return ''
    else: # Si el árbol no está vacío
        mostrar_arbol(arbol.der, cont+1)
        
        for i in range(cont):
            print("   ", end='')
        
        print(arbol.dato)
        mostrar_arbol(arbol.izq, cont+1)

# Función para buscar un elemento en el árbol
def buscar_nodo(arbol: Nodo, n: int):
    if arbol.dato == None:
        return False
    elif arbol.dato == n:
        return True
    elif n < arbol.dato:
        return buscar_nodo(arbol.izq, n)
    else:
        return buscar_nodo(arbol.der, n)

# Función para recorrer el árbol en PreOrden
def pre_orden(arbol: Nodo):
    if arbol.dato == None:
        return ""
    else:
        print("{} - ".format(arbol.dato), end="")
        pre_orden(arbol.izq)
        pre_orden(arbol.der)

# Función para recorrer el árbol en InOrden
def in_orden(arbol: Nodo):
    if arbol.dato == None:
        return ""
    else:
        pre_orden(arbol.izq)
        print("{} - ".format(arbol.dato), end="")
        pre_orden(arbol.der)

# Función para recorrer el árbol en PostOrden
def post_orden(arbol: Nodo):
    if arbol.dato == None:
        return ""
    else:
        pre_orden(arbol.izq)
        pre_orden(arbol.der)
        print("{} - ".format(arbol.dato), end="")

def menu():
    dato: int
    opcion: int = 0
    contador: int = 0
    
    arbol = Nodo()
    
    while opcion != 7:
        print("\t.:MENU:.")
        print("1. Insertar un nuevo nodo.")
        print("2. Mostrar el arbol.")
        print("3. Buscar un elemento en el árbol.")
        print("4. Recorrer el árbol en PreOrden.")
        print("5. Recorrer el árbol en InOrden.")
        print("6. Recorrer el árbol en PostOrden.")
        print("7. Salir.")
        opcion = int(input("Opción: "))
        
        if opcion == 1:
            dato = int(input("\nDigite un número: "))
            arbol = insertar_nodo(arbol, dato)
        elif opcion == 2:
            print("\nMostrando el árbol completo:\n")
            mostrar_arbol(arbol, contador)
            print()
        elif opcion == 3:
            dato = int(input("\nDigite el elemento a buscar: "))
            if buscar_nodo(arbol, dato):
                print("El elemento {} ha sido encontrado en el árbol.".format(dato))
            else:
                print("El elemento {} no se ha encontrado en el árbol.".format(dato))
        elif opcion == 4:
            print("\nRecorrido en PreOrden:")
            pre_orden(arbol)
            print()
        elif opcion == 5:
            print("\nRecorrido en InOrden:")
            in_orden(arbol)
            print()
        elif opcion == 6:
            print("\nRecorrido en PostOrden:")
            post_orden(arbol)
            print()

def main():
    
    menu()

if __name__ == "__main__":
    main()
    
    